### PR TITLE
[CodeQuality] Add ForRepeatedCountToOwnVariableRector

### LINKED ISSUE #1573: [CodeQuality] Looping count to variable with count
```diff
+$itemCount = count($node->args);
-for ($i = $reflectionFunctionLike->getNumberOfParameters(); $i <= count($node->args); $i++) {
+for ($i = $reflectionFunctionLike->getNumberOfParameters(); $i <= $itemCount; $i++) {
     unset($node->args[$i]);
 }
 ```

PHPStorm edit:

```diff
-for ($i = $reflectionFunctionLike->getNumberOfParameters(); $i <= count($node->args); $i++) {
+for ($i = $reflectionFunctionLike->getNumberOfParameters(), $iMax = count($node->args); $i <= $iMax; $i++) {
```

### PR DESCRIPTION
**Pull request by TomasVotruba**
Closes #1573

### CODE REVIEW

File: packages/CodeQuality/src/Rector/For_/ForRepeatedCountToOwnVariableRector.php [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+            $countInCond = $node;
+            $valueName = $this->getName($node->args[0]->value);
+
+            if ($valueName === null) {
+                $countedValueName = self::DEFAULT_VARIABLE_COUNT_NAME;
+            } else {
+                $countedValueName = $valueName . 'Count';
+            }
+
```
Comment (staabm) [inline packages/CodeQuality/src/Rector/For_/ForRepeatedCountToOwnVariableRector.php]: Should the rector make sure that there is no other - already existing variable - in the current scope which might collide with he newly introduced one?

Comment (TomasVotruba) [reply packages/CodeQuality/src/Rector/For_/ForRepeatedCountToOwnVariableRector.php]: Could you provide code example?

How would you name new variable in case of conflict?

Comment (staabm) [reply packages/CodeQuality/src/Rector/For_/ForRepeatedCountToOwnVariableRector.php]:
```php
class SomeClass
 {
     public function run($items)
     {
         $itemsCount = 25;
         for ($i = 5; $i <= count($items); $i++) {
             echo $items[$i];
         }
         If ($itemsCount == 25 ) echo „whoo“;
     }
 }
```

Comment (TomasVotruba) [reply packages/CodeQuality/src/Rector/For_/ForRepeatedCountToOwnVariableRector.php]: I see

Comment (TomasVotruba) [reply packages/CodeQuality/src/Rector/For_/ForRepeatedCountToOwnVariableRector.php]: Covered in https://github.com/rectorphp/rector/pull/2562

Comment (staabm) [reply packages/CodeQuality/src/Rector/For_/ForRepeatedCountToOwnVariableRector.php]: Perfect, thx

File: packages/CodeQuality/tests/Rector/For_/ForRepeatedCountToOwnVariableRector/Fixture/fallback_for_complex.php.inc [hunk 2]
```diff
@@ -0,0 +1,32 @@
+<?php
+
+namespace Rector\CodeQuality\Tests\Rector\For_\ForRepeatedCountToOwnVariableRector\Fixture;
+
+class FallbackForComplex
+{
+    public function run($items, \stdClass $someObject)
+    {
+        for ($i = 5; $i <= count($someObject->getItems() + 10); $i++) {
+            echo $items[$i];
```
Comment (staabm) [inline packages/CodeQuality/tests/Rector/For_/ForRepeatedCountToOwnVariableRector/Fixture/fallback_for_complex.php.inc]: In case the loop body changes the number of elements of the array the `count()` cannot be calculated beforehand .. could happen when the array get re-assigned, changed by reference or elements get added/deleted

Comment (TomasVotruba) [reply packages/CodeQuality/tests/Rector/For_/ForRepeatedCountToOwnVariableRector/Fixture/fallback_for_complex.php.inc]: There are many theoretical edge cases. Not worth covering unless really exist
